elastic/apm-server
# Add experimental profiling collection agent code

## Motivation/summary

Integrate the profiling collection agent code into APM Server, as a technical preview / experiment.

This feature is disabled by default, and can be enabled by setting `apm-server.profiling.enabled=true`. The profiling code will use its own go-elasticsearch client to write to `profiling-*` indices and data streams. Management of these indices / data streams is handled externally to APM Server.

When running under Elastic Agent & Fleet, APM Server's privileges are insufficient to writing to `profiling-*`, so it is necessary to provide an additional API Key for the profiling code to use. This can be injected (e.g. [using fleetctl](https://github.com/elastic/apm-server/tree/main/systemtest/cmd/fleetctl#updating-arbitrary-config)) as `apm-server.profiling.elasticsearch.api_key`.

## Checklist

~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~
~- [ ] Documentation has been updated~

## How to test these changes

1. Run `apm-server -E apm-server.profiling.enabled=true ...`
2. Run `pf-host-agent -disable-tls -collection-agent localhost:8200 -project-id 123 -secret-token abc`
3. Confirm data shows up in Kibana

Also check that apm-server's secret token auth is honoured. It is not currently possibly to configure the host agent to use API Key auth.

## Related issues

https://github.com/elastic/apm-server/issues/9051
